File: Step 06 LinkedList/Step 6.3 Medium Problem of LL/Sort LL 0s 1s 2s/Sort LL 0s 1s 2s.cpp
/*
 
  Node is defined as
  struct Node {
    int data;
    struct Node *next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

*/
class Solution
{
    public:
    //Function to sort a linked list of 0s, 1s and 2s.
    Node* segregate(Node *head) {
        
        // Add code here
        Node *zeros=new Node(-1);
        Node *ones=new Node(-1);
        Node *twos=new Node(-1);
        Node *curr=head;
        Node *headones=ones;
        Node *headtwos=twos;
        Node *headzeros=zeros;
        while(curr!=NULL){
            if(curr->data==0){
                zeros->next=curr;
                zeros=zeros->next;
            }
            if(curr->data==1){
                ones->next=curr;
                ones=ones->next;
            }
            if(curr->data==2){
                twos->next=curr;
                twos=twos->next;
            }
            curr=curr->next;
        }
        zeros->next=(headones->next!=NULL) ? headones->next : headtwos->next;
        ones->next=headtwos->next;
        twos->next=NULL;
        head=headzeros->next;
        return head;
    }
};

File: Step 06 LinkedList/Step 6.3 Medium Problem of LL/Sort LL 0s 1s 2s/Sort LL 0s 1s 2s.java
/****************************************************************

 Following is the class structure of the Node class:

 class Node {
     public int data;
     public Node next;
    
     Node()
     {
         this.data = 0;
         this.next = null;
     }
    
     Node(int data)
     {
         this.data = data;
         this.next = null;
     }
    
     Node(int data, Node next)
     {
         this.data = data;
         this.next = next;
     }
 }

 *****************************************************************/

public class Solution
{
    public static Node sortList(Node head) {
        // Write your code here
        Node zeroHead = new Node(0);
        Node zero = zeroHead;
        Node oneHead = new Node(0);
        Node one = oneHead;
        Node twoHead = new Node(0);
        Node two = twoHead;
        Node temp = head;
        while(temp != null) {
            if (temp.data == 0) {
                zero.next = temp;
                zero = zero.next;
            } else if (temp.data == 1) {
                one.next = temp;
                one = one.next;
            } else {
                two.next = temp;
                two = two.next;
            }
            temp = temp.next;
        }
        two.next = null;
        one.next = twoHead.next;
        zero.next = oneHead.next;
        return zeroHead.next;
    }
}